python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -583,10 +588,28 @@ def my_function():
     resource_defs_keys = set(resource_defs.keys())
     required_resource_keys = bare_required_resource_keys | resource_defs_keys
 
+    asset_out_map: Mapping[str, AssetOut] = {} if outs is None else outs
+
     def inner(fn: Callable[..., Any]) -> AssetsDefinition:
         op_name = name or fn.__name__
         asset_ins = build_asset_ins(fn, ins or {}, deps=_make_asset_keys(upstream_asset_deps))
-        output_tuples_by_asset_key = build_asset_outs(outs)
+
+        if asset_out_map and specs:
+            raise DagsterInvalidDefinitionError("Must specify only outs or assets but not both.")
+        elif specs:
+            output_tuples_by_asset_key = {}
+            for asset_node in specs:
+                # output names are asset keys joined with _
+                output_name = "_".join(asset_node.asset_key.path)
+                output_tuples_by_asset_key[asset_node.asset_key] = (
+                    output_name,
+                    Out(
+                        Nothing,
+                        is_required=not can_subset,
```

Comment:
this is an interesting design choice worthy of scrutiny - the optionality of all outputs is controlled by declaring `can_subset`, a sort of all-or-nothing opt-in.

Reply:
Yeah I think we need to support `multi_asset` with one of the assets being optional, but not subset-able

Reply:
Any thoughts on how that should be expressed? Should it be a property of the `AssetSpec` ?

Reply:
Attaching it to `AssetSpec` feels a bit weird. Seems like it should be a property of the computation.

Reply:
not sure on the name `optional_specs` but cleanest way i can think so far is to have a sibling argument to "specs"
 for the set of assets that are conditionally materialized (maybe it should be `conditional_specs` )

```
    @multi_asset(
        specs=[table_A],
        optional_specs=[table_B],
    )
    def basic_optional(context: AssetExecutionContext):
        yield MaterializeResult(asset_key="table_A")
```

```
    @multi_asset(
        optional_specs=[table_A, table_B],
        can_subset=True, # all specs must be optional if can_subset
    )
    def basic_subset(context: AssetExecutionContext):
        for key in context.selected_asset_keys:
            yield MaterializeResult(asset_key=key)
```

@sryza thoughts?

Reply:
one direction that my thoughts go is whether there will be other orthogonal properties that we'll want to set per-output but don't make sense to put on the `AssetSpec`

can't immediately think of one though

Reply:
@sryza is there an argument for the putting the optional on `AssetSpec`. It could be argued that it is an essential property of an asset is whether or not it's materialization function actually materializes it unconditionally?

Reply:
It's a little sketchy but I don't think super problematic: dagster_type, code_version, and io_manager_key are all related to the materialization function as well.

Reply:
well specs drop `io_manager_key` and `dagster_type`, so `code_version` is the last odd ball but i think can be improved with rename/docstring cleanup

Reply:
Wait so what happens to `io_manager_key`?

Reply:
The current thinking is using specs is for un-managed IO so no interactions with IO managers and no need to set the key. Are we missing something?

Reply:
I see. I guess I had been assuming that we'd basically deprecate `AssetOut`, but it sounds like we'll be supporting both? Not so bad, just different from what I was expecting.

Reply:
Got it - makes sense. This makes me continue to dream of having some sort of storage concept that applies to assets even if we're not managing the output.